File: 1514. Path with Maximum Probability.py
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
        
        graph = [[] for _ in range(n)]
        for i, (a, b) in enumerate(edges):
            prob = succProb[i]
            graph[a].append((b, prob))
            graph[b].append((a, prob))
        
        dist = [0] * n
        dist[start] = 1.0
        
        pq = [(-1.0, start)]
        heapq.heapify(pq)
        
        while pq:
            prob, node = heapq.heappop(pq)
            prob *= -1.0
            
            if node == end:
                return prob
            
            if prob < dist[node]:
                continue
            
            for neighbor, edge_prob in graph[node]:
                new_prob = prob * edge_prob
                if new_prob > dist[neighbor]:
                    dist[neighbor] = new_prob
                    heapq.heappush(pq, (-new_prob, neighbor))
        
        return 0.0

File: 1514. Path with Maximum Probability.java
class Solution {
    class Pair{
        int node;
        double probability;
        Pair(int node,double probability){
            this.node = node;
            this.probability = probability;
        }
    }
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
        for(int i=0;i<n;i++){
            adj.add(new ArrayList<Pair>());
        }
        int ind = 0;
        for(int[] edge:edges){
            adj.get(edge[0]).add(new Pair(edge[1],succProb[ind]));
            adj.get(edge[1]).add(new Pair(edge[0],succProb[ind]));
            ind++;
        }

        return Diskistras(adj,start_node,end_node,n);
    }

    public double Diskistras(ArrayList<ArrayList<Pair>> adj,int start,int end,int n){
        // heap 
        PriorityQueue<Pair> pq = new PriorityQueue<>((x,y)->Double.compare(y.probability,x.probability));

        double[] probabilites = new double[n];
        probabilites[start] = 1d;

        pq.add(new Pair(start,1d));

        while(!pq.isEmpty()){
            Pair pair = pq.poll();

            int  currNode = pair.node;
            double currProbability = pair.probability;

            if(currNode == end){
                return probabilites[currNode];
            }

            for(Pair item : adj.get(currNode)){
                int adjNode = item.node;
                double adjProb = item.probability;
                if( (currProbability*adjProb) > probabilites[adjNode] ){
                    probabilites[adjNode] = currProbability*adjProb;
                    pq.add(new Pair(adjNode,probabilites[adjNode]));
                }
            }
        }
        return  probabilites[end];

    }
}
